flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1058,6 +1092,30 @@ public void handleError(Throwable throwable) {
 		}
 	}
 
+	/**
+	 * The heartbeat listener for JobManager and ResourceManager, they can be distinguished by ResourceID
+	 * and trigger different processes.
+	 */
+	private final class JMRMHeartbeatListener implements HeartbeatListener {
+
+		JMRMHeartbeatListener() {
+		}
+
+		@Override
+		public void notifyHeartbeatTimeout(final ResourceID resourceID) {
+			log.info("Notify heartbeat timeout for resourceID {}", resourceID);
```

Comment:
Yes, it actually should trigger some actions with timeout. Currently I did not submit this part because I think it is related with failure detection logic  and supposed to submit in another PR. To make the heartbeat mechanism complete, I will add this part in the following modifications.